Step3

nocomet opened the pull request:
안녕하세요~ step3 미션을 수행했습니다.

과제를 진행하면서 비즈니스로직을 service layer에 작성하는 습관이 있었는데요

이번 리팩토링 과제를 수행하면서 뭔가 큰 깨달음을 얻은 것 같습니다 :)

감사합니다 피드백주시면 반영하도록 하겠습니다.

star-no committed:
docs: README 작성

star-no committed:
feature: Question, Answer 삭제 기능 구현

Junhong-Kim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -93,6 +94,22 @@ public void setDeleted(boolean deleted) {
         this.deleted = deleted;
     }
 
+    void canDelete(User loginUser) throws CannotDeleteException {
```
클래스 내부에서만 쓰이는 메서드는 접근 제어자를 `private`으로 설정하면 좋을 것 같아요.

Junhong-Kim commented on src/main/java/qna/domain/Question.java:
```diff
@@ -87,6 +93,27 @@ public void setDeleted(boolean deleted) {
+        canDelete(loginUser);
+
+        DeleteHistories deleteHistories = new DeleteHistories();
+        deleteHistories.add(DeleteHistory.byQuestion(id, loginUser));
+        deleteHistories.add(answers.delete(loginUser));
```
답변이 모두 삭제되어야만 질문을 삭제할 수 있으므로 삭제 순서를 바꾸는게 어떨까요?

nocomet replied:
기존에 미리 작성된 `QnaServiceTest` 의 
```
private void verifyDeleteHistories() {
    List<DeleteHistory> deleteHistories = Arrays.asList(
            new DeleteHistory(ContentType.QUESTION, question.getId(), question.getWriter(), LocalDateTime.now()),
            new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now())
    );
    verify(deleteHistoryService).saveAll(deleteHistories);
}
```
이 코드에서 `deleteHistories` 안에 `QUESTION DeleteHistory` 부터 array에 들어가서 테스트 케이스에 실패하는데요, 이건 테케를 바꾸면 될지,, 아니면 삭제한 결과를 add하는  시점을 다르게 해서 해결할지 질문 드립니다~

Junhong-Kim replied:
현재 수정해주신 내용으로도 괜찮을 것 같아요 ☺️

Junhong-Kim commented on src/main/java/qna/domain/Question.java:
```diff
@@ -87,6 +93,27 @@ public void setDeleted(boolean deleted) {
+        deleteHistories.add(DeleteHistory.byQuestion(id, loginUser));
+        deleteHistories.add(answers.delete(loginUser));
+
+        this.deleted = true;
```
질문을 삭제 처리하는 코드와 질문 삭제 히스토리를 코드를 하나의 메서드로 추출하면 좋을 것 같아요.

nocomet replied:
이 질문도 제가 의도를 잘 이해하지 못했습니다,,

`질문` 삭제하는 코드와 `답변` 삭제하는 코드를 하나의 메소드로 추출하라는 말씀이실까요?

Junhong-Kim replied:
`Answer` 클래스의 `delete(User loginUser)`를 참고해보시면 될 것 같아요!
다음 로직들도 하나의 행위로 볼 수 있으므로 유사하게 처리하면 좋을 것 같습니다.
```java
this.deleted = true;
deleteHistories.add(DeleteHistory.byQuestion(id, loginUser));
```

nocomet replied:
아! 가독성을 위해 메소드 추출을 하란 의미셨을까요? 👍👍👍👍

Junhong-Kim replied:
넵, 맞습니다!

Junhong-Kim commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,41 @@
+
+    private List<Answer> undeletedAnswers() {
+        return answers.stream()
+                .filter(answer -> !answer.isDeleted())
```
`!` 사용을 지양하고 메서드로 추출해서 가독성있는 코드로 사용하면 좋을 것 같아요.

nocomet replied:
오.. 피드백 주신 내용을 잘 이해하지 못했어요 ㅜㅜ
1. `!` 을 사용하지 않으면 삭제되지 않은 것들을 모으는 조건을 어떻게 하면 될까요?
2. 가독성있는 메소드 추출이 filter 안에 있는 `answer -> !answer.isDeleted()` 부분을 추출하란 말씀이실까요?

Junhong-Kim replied:
`isNotDeleted()` 메서드로 잘 처리해주셨네요 👍

Junhong-Kim commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,38 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class AnswersTest {
```
`Answers` 외에 도메인에 해당하는 테스트 작성이 안되어있네요.
`Answers` 도메인도 기능에 대한 테스트 코드 작성이 부족한 것 같아요.
다른 도메인 클래스들의 테스트 코드도 작성해주세요 ☺️

Junhong-Kim reviewed (changes requested):
3단계 미션 잘 구현해주셨습니다 👍
리뷰 내용 확인하시고 다시 요청해주시면 감사드립니다!

star-no committed:
refact: 코드리뷰 반영

1. Answer, Question, DeleteHistories TestCase 추가
2. Answers TestCase 보강
3. 패키지 접근 범위인 함수를 private 으로 변경

Junhong-Kim commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,44 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+
 public class QuestionTest {
```
질문에 다른 사용자 답변이 삭제되었을 때 질문을 삭제하는 케이스와
질문에 다른 사용자 답변이 삭제되지 않았을 때 질문을 삭제하는 케이스에 대해 테스트가 있으면 좋을 것 같습니다.

Junhong-Kim commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,45 @@
+
+    @DisplayName("delete 성공")
+    @Test
+    void deleteTest01() {
```
테스트 코드 메서드명으로 숫자를 사용하고 있으신데요.
영어로 메서드명을 지으려는 것보다는 테스트 코드 메서드에서 예외적으로 한글을 사용하는 것도 방법입니다 ☺️

Junhong-Kim commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -35,4 +35,26 @@ public void deleteTest02() {
+
+        DeleteHistories delete = answers.delete(UserTest.JAVAJIGI);
+
+        assertThat(delete).isNotNull();
+        assertThat(delete.getList())
```
`assertAll()`을 활용한 검증 방법에 대해 알아보고 적용해보면 좋을 것 같아요.

nocomet replied:
` assertThat(delete).isNotNull();`
이게 검증이 되어야 밑에 assert 들도 의미가 있기 때문에 별도로 작성했습니다.

delete 객체가 null이면 그 아래 테스트는 npe로 돌아가는데 의미가 없다고 생각했거든요.

그런데 음, 같이 코드를 읽는 coworker에겐 혼란이 올 수도 있겠단 생각이 드네요
이런 경우에도 Assertions.assertAll 로 처리하도록 습관을 가져야겠습니다 :)

Junhong-Kim commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,44 @@
+
+    @DisplayName("delete 성공")
+    @Test
+    void deleteTest01() {
```
전반적으로 테스트 코드에 `given-when-then` 패턴을 적용해보면 좋을 것 같아요.
참고할만한 [링크](https://brunch.co.kr/@springboot/292)도 공유드립니다 ☺️

Junhong-Kim reviewed (changes requested):
피드백 잘 반영해주셨는데 좀 더 수정해볼 만한 내용 코멘트 남겼습니다!
그리고 PR로 질문 남겨주신 사항들은 대댓글로 답변드렸습니다.
리뷰 내용 확인하시고 다시 요청해주시면 감사드립니다. ☺️

star-no committed:
refact: Question.delete 함수코드를 가독성을 위해 메소드 추출

star-no committed:
refact: QuestionTest testcast 보강

Junhong-Kim reviewed (approved):
마지막 피드백까지 잘 반영해주셨습니다 👍
리뷰 반영하시느라 정말 고생 많으셨어요!

이번 미션은 여기서 마무리하겠습니다.
다음 미션도 화이팅입니다! ☺️
